### PR TITLE
Upgraded Checkpoint Module to Save Models to disk

### PR DESCRIPTION
**Pull request by danich1**
This PR is designed as a patch work for issue #88. Originally metal saved the best model during training in memory and I added the functionality to save models to the disk instead. Note that I am saving the model's state dict instead of the whole model object itself. I also allowed for the model to resume training at a specified epoch, so users don't have to worry about losing all their work. Let me know what you guys think. 

Also fixed a small bug in the multitask code where it should check for the size of K, not if K exists.

### CODE REVIEW

File: metal/classifier.py [hunk 1]
```diff
@@ -182,8 +212,28 @@ def _train_model(self, train_data, loss_fn, dev_data=None, log_writer=None):
                 train_config["checkpoint_config"]
             )
 
+        # which iteration to start with
+        start_iteration = 0
+
+        if restore_state:
```
Comment (bhancock8) [inline metal/classifier.py]: Can we move this loading functionality to helper? _train_model() is already uncomfortably long.

File: metal/multitask/mt_end_model.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def __init__(
         MTClassifier.__init__(self, K, config)
 
         if task_graph is None:
-            if K is None:
+            if len(K) == 0:
```
Comment (bhancock8) [inline metal/multitask/mt_end_model.py]: Nice catch!

File: metal/utils.py [hunk 3]
```diff
@@ -35,7 +36,12 @@ def __len__(self):
 
 class Checkpointer(object):
     def __init__(
-        self, model_class, checkpoint_min=-1, checkpoint_runway=0, verbose=True
+        self,
+        model_class,
+        checkpoint_min=-1,
+        checkpoint_runway=0,
+        verbose=True,
+        checkpoint_destination="checkpoints",
```
Comment (bhancock8) [inline metal/utils.py]: Add docstring under Args

File: metal/utils.py [hunk 4]
```diff
@@ -49,27 +55,54 @@ def __init__(
         self.best_iteration = None
         self.best_score = checkpoint_min
         self.checkpoint_runway = checkpoint_runway
+        self.checkpoint_destination = checkpoint_destination
         self.verbose = verbose
+        self.state = dict({})
```
Comment (bhancock8) [inline metal/utils.py]: why not just {}?

File: metal/utils.py [hunk 5]
```diff
@@ -49,27 +55,54 @@ def __init__(
...
             print(
                 f"No checkpoints will be saved in the first "
                 f"checkpoint_runway={checkpoint_runway} iterations."
             )
 
-    def checkpoint(self, model, iteration, score):
+    def checkpoint(self, model, iteration, score, optimizer, lr_scheduler):
         if iteration >= self.checkpoint_runway:
+            self.state["epoch"] = iteration
+            self.state["model"] = (model.state_dict(),)
```
Comment (bhancock8) [inline metal/utils.py]: Why do these need to be tuples?

**Comment by bhancock8** [comment]
(Also, sorry for the deluge of individual comments; I was trying out reviewing PRs directly in VSCode with their new extension and apparently the feature isn't in yet to let you group all your comments in a single review). :)

**Comment by ajratner** [comment]
@danich1 Thanks so much for this!!!  Otherwise, all I have to say is: whatever @bhancock8 said :)

File: metal/utils.py (hunk 5 above)
Comment (danich1) [reply metal/utils.py]: No reason. I think it was a quick fix for a bug I ran into previously and I forgot to change it back.

**Comment by bhancock8** [comment]
Awesome! I'll take a close look once it's passing all tests. It looks from the travis-ci output that two of our notebook tutorials (Basics.ipynb and Multitask.ipynb) break because of changes in this PR. Take a look and see what line (I suspect it will probably just be a simple signature change in one line) needs to be updated in each to pass the tests.

**Comment by bhancock8** [comment]
lgtm! Thanks @danich1!

**Pull request merged**